### DISCUSSION TITLE
I need some help with optimizing my script

### DISCUSSION THREAD

**Discussion by imageck** [ImageMagick/ImageMagick#4935] [Help] [answered]
Hello!

I need some help with optimizing my script. Here's an excerpt concerning the usage of ImageMagick:

```
xfce_backdrop_get_quad_average()
{
# samples four corners
# c1-----c2
# |       |
# c3-----c4

	corners=$(convert "${backdrop}" -crop 2x2@ +repage +adjoin -sample 1x1\! rgb:- |
	od --address-radix=n --format=uC --width=12)

	read -r \
		corner1_red corner1_green corner1_blue \
		corner2_red corner2_green corner2_blue \
		corner3_red corner3_green corner3_blue \
		corner4_red corner4_green corner4_blue \
	<<-eof
		${corners}
	eof

	center=$(convert "${backdrop}" -gravity Center \
		-crop $((width/2))x$((height/2))+0+0 +repage -sample 1x1\! rgb:- |
	od --address-radix=n --format=uC --width=12)

	read -r center_red center_green center_blue <<-eof
		$center
	eof
	xfce_backdrop_get_average_value
}
```

The rest of the script is a work-in-progress but this specific function works as I intend it to. However, I see two problems here:
1) The `convert` program is invoked twice, separately. I would like to do the whole processing in one go. What I basically want is to take a single image (the full pathname is stored in `$backdrop`), divide it into four equal regions/quadrants (`2x2@` does precisely that) plus the central region/quadrant with the same dimensions (width/2 x height/2). Essentially, NW, NE, SW, SE and central regions are extracted from the image, and then scaled down to 1x1, and finally RGB values of each 1x1 image is sent to stdout for further processing and action.
2) As you can see, it then reformats their raw RGB values to unsigned decimals using GNU od. This involves creating another subshell forking another process, which is suboptimal. Are there any other special output formats or `-format` strings that I can use to print out only RGB values? Currently, with those two `convert` executions I get two lines of output, which are processed later independently:
```
120 38 100 175 56 53 110 29 99 154 47 69 # RGB values of each corner
139 41 79 # RGB values of center
```
It's fine if a `-format` string contains some garbage, I can just strip that out with parameter expansion. I just want the most efficient, fast, the least resource-intensive method.

Each corner is then assigned a weight and some arithmetic is done to get a single RGB value:
`average="${average_red},${average_green},${average_blue}"`
Is it possible to do that in ImageMagick as well? `-fx` seems like a good candidate but the documentation says it's very slow, and also I'm not familiar with it.

Thanks.

**Comment by fmw42** [comment, marked as answer]
Here is one way to do that. I write the image into MPR: in-memory format and then use it over again for each gravity setting to get the R,G,B colors using -fx and combine them all into one array.  Then I take every third one to get the averages.  (Note, -fx is much faster now in IM 7, but this is not slow for what we are doing here)

![lena](https://user-images.githubusercontent.com/13458218/158026761-f12816b2-cf6c-49e9-a3cf-22b322dfc999.png)

```
infile="lena.png"
colorArr=(`convert "$infile" -write mpr:img +delete \
\( mpr:img -gravity northwest -crop 25x25%+0+0 +repage -scale 1x1! \
    -format "%[fx:round(255*u.r)] %[fx:round(255*u.g)] %[fx:round(255*u.b)] " -write info: \) \
\( mpr:img -gravity northeast -crop 25x25%+0+0 +repage -scale 1x1! \
    -format "%[fx:round(255*u.r)] %[fx:round(255*u.g)] %[fx:round(255*u.b)] " -write info: \) \
\( mpr:img -gravity southeast -crop 25x25%+0+0 +repage -scale 1x1! \
    -format "%[fx:round(255*u.r)] %[fx:round(255*u.g)] %[fx:round(255*u.b)] " -write info: \) \
\( mpr:img -gravity southwest -crop 25x25%+0+0 +repage -scale 1x1! \
    -format "%[fx:round(255*u.r)] %[fx:round(255*u.g)] %[fx:round(255*u.b)] " -write info: \) \
\( mpr:img -gravity center -crop 25x25%+0+0 +repage -scale 1x1! \
    -format "%[fx:round(255*u.r)] %[fx:round(255*u.g)] %[fx:round(255*u.b)]\n" -write info: \) \
null:`)

red_ave=`echo "scale=6; (${colorArr[0]}+${colorArr[3]}+${colorArr[6]}+${colorArr[9]}+${colorArr[12]})/5" | bc`
green_ave=`echo "scale=6; (${colorArr[1]}+${colorArr[4]}+${colorArr[7]}+${colorArr[10]}+${colorArr[13]})/5" | bc`
blue_ave=`echo "scale=6; (${colorArr[2]}+${colorArr[5]}+${colorArr[8]}+${colorArr[11]}+${colorArr[14]})/5" | bc`

echo "red_ave=$red_ave; green_ave=$green_ave; blue_ave=$blue_ave;"
red_ave=173.800000; green_ave=92.600000; blue_ave=99.400000;

```

**Reply by imageck** [reply]
Woah! This command, compared to the one in my script, cuts the execution time by almost a half: 
```
real	0m0.582s           real	0m0.335s
user	0m0.505s    vs     user	0m0.363s
sys	0m0.206s           sys	0m0.083s
```

**Reply by imageck** [reply]
@fmw42 thanks for sharing this one! 

One question, however, remains: what if I wanted RGB values in the [0..1] range? I realized that I need them in this format in the end. I see that ` -format "%[fx:r] %[fx:g] %[fx:b]\n"` does this. Any other expression? Can it be done with the`rgb()` function?

**Reply by fmw42** [reply]
u.r, u.g, u.b are the R,G,B values in the range 0 to 1.  The u means the first input image.  See https://imagemagick.org/script/fx.php. So just remove the multiplication by 255.

```
infile="lena.png"
colorArr=(`convert "$infile" -write mpr:img +delete \
\( mpr:img -gravity northwest -crop 25x25%+0+0 +repage -scale 1x1! \
    -format "%[fx:u.r] %[fx:u.g] %[fx:u.b] " -write info: \) \
\( mpr:img -gravity northeast -crop 25x25%+0+0 +repage -scale 1x1! \
    -format "%[fx:u.r] %[fx:u.g] %[fx:u.b] " -write info: \) \
\( mpr:img -gravity southeast -crop 25x25%+0+0 +repage -scale 1x1! \
    -format "%[fx:u.r] %[fx:u.g] %[fx:u.b] " -write info: \) \
\( mpr:img -gravity southwest -crop 25x25%+0+0 +repage -scale 1x1! \
    -format "%[fx:u.r] %[fx:u.g] %[fx:u.b] " -write info: \) \
\( mpr:img -gravity center -crop 25x25%+0+0 +repage -scale 1x1! \
    -format "%[fx:u.r] %[fx:u.g] %[fx:u.b]\n" -write info: \) \
null:`)

red_ave=`echo "scale=6; (${colorArr[0]}+${colorArr[3]}+${colorArr[6]}+${colorArr[9]}+${colorArr[12]})/5" | bc`
green_ave=`echo "scale=6; (${colorArr[1]}+${colorArr[4]}+${colorArr[7]}+${colorArr[10]}+${colorArr[13]})/5" | bc`
blue_ave=`echo "scale=6; (${colorArr[2]}+${colorArr[5]}+${colorArr[8]}+${colorArr[11]}+${colorArr[14]})/5" | bc`

echo "red_ave=$red_ave; green_ave=$green_ave; blue_ave=$blue_ave;"
red_ave=.681535; green_ave=.362938; blue_ave=.390215;

```

**Comment by snibgo** [comment]
What version of IM? If it isn't recent, I suggest you upgrade to the current v7 version. Then use "magick" instead of "convert".

What is your script language? Bash?

Instead of writing to "rgb:" I suggest writing to "ftxt:", so you don't need "od". Then you can:
```
magick infile.ext -define "ftxt:format=\c " ftxt:-
```
Note there is a space after "\c", so all the pixel values will be on one line.

You can call magick once, and read your input once, and write multiple outputs. For example:
```
magick "${backdrop}" \
  \( -crop 2x2@ +repage +adjoin -sample 1x1\! -define "ftxt:format=\c " ftxt:- +delete \) \
  -gravity Center \
  -crop $((width/2))x$((height/2))+0+0 +repage -sample 1x1\! ftxt:-
```
"ftxt:" is documented at [fmttxt:](http://im.snibgo.com/fmttxt.htm), although on that page it is called "fmttxt:".

Note that it writes decimal values in the nominal range 0 to 1, or 0 to 100, or 0 to QuantumLevel. If you want 0 to 255, you can multiply the values before writing them (for example "-evaluate Multiply %[fx:255/QuantumLevel]").

"-fx" used to be slow. In very recent versions of IM, it is now fast.

**Reply by imageck** [reply]
Oops. Sorry, I forgot to mention the most important details.
> What version of IM? If it isn't recent, I suggest you upgrade to the current v7 version. Then use "magick" instead of "convert".

It's 6.9.10.23 Q16 (on Linux x86_64 Xubuntu 20.04)
> What is your script language? Bash?

POSIX shell script.

I appreciate your awesome suggestions! Let me try them.

**Reply by imageck** [reply]
> "ftxt:" is documented at [fmttxt:](http://im.snibgo.com/fmttxt.htm), although on that page it is called "fmttxt:".

Unfortunately, this format is not supported, or at least not listed by this version of ImageMagick.
